src/ecc_key.c

```diff
@@ -0,0 +1,1222 @@
+	ec_key->key_set = 0;
+	ec_key->pubkey_set = 0;
+
+	ec_key->curve = 0;
+	ec_key->curve_set = 0;
+
+	ec_key->flags = 0;
+	ec_key->flags_set = 0;
+
+	ec_key->type = 0;
+	ec_key->type_set = 0;
+
+	memset(ec_key->mkvp, 0, sizeof(ec_key->mkvp));
+	ec_key->mkvplen = 0;
+	ec_key->mkvp_set = 0;
+
+	free(ec_key->apqns);
+	ec_key->apqns = NULL;
+	ec_key->napqns = 0;
+	ec_key->apqns_set = 0;
+
+	ec_key->refcount = 1;
+}
+
+int ec_key_clr2sec(struct zpc_ec_key *ec_key, unsigned int flags,
+			const unsigned char *pubkey, unsigned int publen,
+			const unsigned char *privkey, unsigned int privlen)
+{
+	target_t target;
+	int rv, rc = ZPC_ERROR_APQNSNOTSET;
+	size_t i;
+
+	switch (ec_key->type) {
+	case ZPC_EC_KEY_TYPE_CCA:
+		rv = pthread_mutex_lock(&ccalock);
+		assert(rv == 0);
+		rc = ec_key_clr2sec_cca(&cca, ec_key->curve, flags,
+							(unsigned char *)&ec_key->cur.sec,
+							&ec_key->cur.seclen,
+							pubkey, publen, privkey, privlen, true);
+		rv = pthread_mutex_unlock(&ccalock);
+		assert(rv == 0);
+		if (rc != 0)
+			rc = ZPC_ERROR_EC_KEY_PARTS_INCONSISTENT;
+		return rc;
+	case ZPC_EC_KEY_TYPE_EP11:
+		rv = pthread_mutex_lock(&ep11lock);
+		assert(rv == 0);
+		for (i = 0; i < ec_key->napqns; i++) {
+			rc = get_ep11_target_for_apqn(&ep11, ec_key->apqns[i].card,
+					ec_key->apqns[i].domain, &target, true);
+			if (rc)
+				continue;
+			rc = ec_key_clr2sec_ep11(&ep11, ec_key->curve, flags,
+							(unsigned char *)&ec_key->cur.sec,
+							&ec_key->cur.seclen, pubkey, publen,
+							privkey, privlen, target);
+			free_ep11_target_for_apqn(&ep11, target);
+			if (rc == 0)
+				break;
+		}
+		rv = pthread_mutex_unlock(&ep11lock);
+		assert(rv == 0);
+		if (rc != 0)
+			rc = ZPC_ERROR_EC_KEY_PARTS_INCONSISTENT;
+		return rc;
+		break;
+	default:
+		return ZPC_ERROR_KEYTYPE;
+	}
+}
+
+/*
+ * (Re)derive protected key from a secure key.
+ * Caller must hold ec_key's wr lock.
+ */
+int ec_key_sec2prot(struct zpc_ec_key *ec_key, enum ec_key_sec sec)
+{
+	struct ep11kblob_header *ep11hdr;
+	struct pkey_kblob2pkey3 io;
+	struct ec_key *key = NULL;
+	unsigned char buf[1000];
+	int buflen = sizeof(buf);
+	int rc;
+
+	assert(sec == EC_KEY_SEC_OLD || sec == EC_KEY_SEC_CUR);
+
+	if (sec == EC_KEY_SEC_CUR)
+		key = &ec_key->cur;
+	else if (sec == EC_KEY_SEC_OLD)
+		key = &ec_key->old;
+	assert(key != NULL);
+
+	/* Prepend a header for ep11 key blobs */
+	memset(buf, 0, buflen);
+	if (ec_key->type == ZPC_EC_KEY_TYPE_EP11) {
+		ep11hdr = (struct ep11kblob_header *)&buf;
+		ep11hdr->len = sizeof(struct ep11kblob_header) + key->seclen;
+		ep11hdr->version = PKEY_TYPE_EP11_ECC;
+		ep11hdr->bitlen = 0;
+		memcpy(buf + sizeof(struct ep11kblob_header), key->sec, key->seclen);
+		buflen = sizeof(struct ep11kblob_header) + key->seclen;
+
+		/* The ioctl wants the session id field at the beginning of the secure
+		 * key blob to be nulls, so clear it here in the temp buf. */
+		memset(buf + sizeof(struct ep11kblob_header), 0, 32);
```

Comment:
hmm, really? What if the blob is a session-bound key? Then the session intentionally contains a session.
Note that for ep11 keys with header, you must leave the session field as is, but for ep11 keys where the session field is used as header, you must zeroize the session field.

Reply:
Fixed. Code has been reworked.